File: src/python/vimure/test/test_model.py
from typing import Any
import pytest
import logging

import numpy as np
import pandas as pd
import vimure as vm

from sklearn.metrics import f1_score

from . import karnataka_edgelist_vil1_money, suppress_stdout_stderr

logger = logging.getLogger("vm.test.test_model")

def check_final_parameters(model, net_obj):

    """Latent Variables - Parameter rho"""
    assert model.rho is not None
    assert type(model.rho) == np.ndarray
    assert model.rho.shape == (net_obj.L, net_obj.N, net_obj.N, net_obj.K)
    assert model.rho.sum() != 0, "Model returned a trivial solution"

    """Latent Variables - Parameter gamma - shape & rate"""
    assert model.gamma_shp is not None
    assert type(model.gamma_shp) == np.ndarray
    assert model.gamma_shp.shape == (net_obj.L, net_obj.M)
    assert model.gamma_shp.sum() > 0

    assert model.gamma_rte is not None
    assert type(model.gamma_rte) == np.ndarray
    assert model.gamma_rte.shape == (net_obj.L, net_obj.M)
    assert model.gamma_rte.sum() > 0

    """Latent Variables - Parameter phi - shape & rate"""
    assert model.phi_shp is not None
    assert type(model.phi_shp) == np.ndarray
    assert model.phi_shp.shape == (net_obj.L, net_obj.K)
    assert model.phi_shp.sum() > 0

    assert model.phi_rte is not None
    assert type(model.phi_rte) == np.ndarray
    assert model.phi_rte.shape == (net_obj.L, net_obj.K)
    assert model.phi_rte.sum() > 0

    """Latent Variables - Parameter nu - shape & rate"""
    assert model.nu_shp is not None
    assert type(model.nu_shp) == np.float64
    assert model.nu_shp >= 0

    assert model.nu_rte is not None
    assert type(model.nu_rte) == np.float64
    assert model.nu_rte >= 0

class TestVimureWithRandomNetworks:
    """
    Tests of the VimureModel class
    """

    def test_vimure_model_with_standard_sbm(self):
        """
        Tests the VimureModel class with a known synthetic network
        """

        logger.debug("Generating synthetic network (Standard SBM) for testing")
        gt_network = vm.synthetic.StandardSBM(
            N=20,
            M=20,
            L=1,
            K=3,
            C=2,
            avg_degree=2,
            sparsify=False,
        )

        logger.debug("Generating observed X from given Y")
        gt_network._build_X(flag_self_reporter=True)

        logger.debug("Starting VimureModel")
        model = vm.model.VimureModel()
        with pytest.warns(None) as record:
            with suppress_stdout_stderr():
                model.fit(gt_network.X, K=gt_network.K, R=gt_network.R)

        check_final_parameters(model, gt_network)

    def test_vimure_model_issues_appropriate_warnings(self):
        """
        Tests the VimureModel class with a known synthetic network
        """

        logger.debug("Generating synthetic network (Standard SBM) for testing")
        gt_network = vm.synthetic.StandardSBM(
            N=20,
            M=20,
            L=1,
            K=3,
            C=2,
            avg_degree=2,
            sparsify=False,
        )

        logger.debug("Generating observed X from given Y")
        gt_network._build_X(flag_self_reporter=True)

        logger.debug("Starting VimureModel without providing R")
        model = vm.model.VimureModel()
        with suppress_stdout_stderr():
            with pytest.warns(UserWarning):
                model.fit(gt_network.X, K=gt_network.K)

        logger.debug("Starting VimureModel without providing K")
        model = vm.model.VimureModel()
        with suppress_stdout_stderr():
            with pytest.warns(UserWarning):
                model.fit(gt_network.X, R=gt_network.R)

    def test_vimure_model_extreme_scenarios_over_reporting(self):
        """
        Ensures that the model achieves F1-score near 1.0 on a known
            extreme scenario (10% exaggerators over-reporting)
        """

        """
        GENERATE CUSTOM THETA
        """

        eta = 0.2
        theta_ratio = 0.1
        seed = 25
        K = 2

        with pytest.warns(None) as record:
            gt_network = vm.synthetic.Multitensor(
                N=100,
                M=100,
                L=1,
                C=2,
                K=K,
                avg_degree=5,
                sparsify=True,
                seed=seed,  # seed=25 works well with default values
                ExpM=None,
                eta=eta,
            )

            theta = vm.synthetic.build_custom_theta(
                gt_network=gt_network,
                theta_ratio=theta_ratio,
                exaggeration_type="over",
                seed=seed,
            )

            LAMBDA_0 = 0.01
            LAMBDA_DIFF = 0.99
            gt_network._build_X(
                mutuality=eta,  # Same as planted mutuality
                theta=theta,  # Pre-defined theta
                cutoff_X=False,
                lambda_diff=LAMBDA_DIFF,
                flag_self_reporter=True,
                seed=seed,
                verbose=False,
            )

        lambda_k_GT = np.array([[LAMBDA_0, LAMBDA_0 + LAMBDA_DIFF]])
        beta_lambda = 10000 * np.ones((lambda_k_GT.shape))
        alpha_lambda = lambda_k_GT * beta_lambda

        model = vm.model.VimureModel(mutuality=True, verbose=False)
        with pytest.warns(None) as record:
            with suppress_stdout_stderr():
                model.fit(
                    gt_network.X,
                    K=K,
                    seed=seed,
                    theta_prior=(0.1, 0.1),
                    eta_prior=(0.5, 1),
                    alpha_lambda=alpha_lambda,
                    beta_lambda=beta_lambda,
                    num_realisations=2,
                    max_iter=21,
                    R=gt_network.R,
                )

        Y_true = gt_network.Y.toarray()[0].flatten()
        Y_rec = vm.utils.apply_rho_threshold(model, threshold=0.5)[0].flatten()

        assert np.allclose(f1_score(Y_true, Y_rec), 0.92, atol=1e-2)

    def test_vimure_model_extreme_scenarios_over_reporting_dense(self):
        """
        Ensures that the model achieves F1-score near 1.0 on a known
            extreme scenario (10% exaggerators over-reporting)
        """

        """
        GENERATE CUSTOM THETA
        """

        eta = 0.2
        theta_ratio = 0.1
        seed = 25
        K = 2

        with pytest.warns(None) as record:
            gt_network = vm.synthetic.Multitensor(
                N=100,
                M=100,
                L=1,
                C=2,
                K=K,
                avg_degree=5,
                sparsify=True,
                seed=seed,  # seed=25 works well with default values
                ExpM=None,
                eta=eta,
            )

            theta = vm.synthetic.build_custom_theta(
                gt_network=gt_network,
                theta_ratio=theta_ratio,
                exaggeration_type="over",
                seed=seed,
            )

            LAMBDA_0 = 0.01
            LAMBDA_DIFF = 0.99
            gt_network._build_X(
                mutuality=eta,  # Same as planted mutuality
                theta=theta,  # Pre-defined theta
                cutoff_X=False,
                lambda_diff=LAMBDA_DIFF,
                flag_self_reporter=True,
                seed=seed,
                verbose=False,
            )

        lambda_k_GT = np.array([[LAMBDA_0, LAMBDA_0 + LAMBDA_DIFF]])
        beta_lambda = 10000 * np.ones((lambda_k_GT.shape))
        alpha_lambda = lambda_k_GT * beta_lambda

        model = vm.model.VimureModel(mutuality=True, verbose=False)
        with pytest.warns(None) as record:
            with suppress_stdout_stderr():
                model.fit(
                    gt_network.X.toarray(),
                    K=K,
                    seed=seed,
                    theta_prior=(0.1, 0.1),
                    eta_prior=(0.5, 1),
                    alpha_lambda=alpha_lambda,
                    beta_lambda=beta_lambda,
                    num_realisations=2,
                    max_iter=21,
                    R=gt_network.R.toarray(),
                )

        Y_true = gt_network.Y.toarray()[0].flatten()
        Y_rec = vm.utils.apply_rho_threshold(model, threshold=0.5)[0].flatten()

        assert np.allclose(f1_score(Y_true, Y_rec), 0.92, atol=1e-2)

    def test_vimure_model_extreme_scenarios_under_reporting(self):
        """
        Ensures that the model achieves F1-score near 1.0 on a known
            extreme scenario (10% exaggerators under-reporting)
        """

        """
        GENERATE CUSTOM THETA
        """

        eta = 0.2
        theta_ratio = 0.1
        seed = 25
        K = 2

        with pytest.warns(None) as record:
            gt_network = vm.synthetic.Multitensor(
                N=100,
                M=100,
                L=1,
                C=2,
                K=K,
                avg_degree=5,
                sparsify=True,
                seed=seed,  # seed=25 works well with default values
                ExpM=None,
                eta=eta,
            )

            theta = vm.synthetic.build_custom_theta(
                gt_network=gt_network,
                theta_ratio=theta_ratio,
                exaggeration_type="under",
                seed=seed,
            )

            LAMBDA_0 = 0.01
            LAMBDA_DIFF = 0.99
            gt_network._build_X(
                mutuality=eta,  # Same as planted mutuality
                theta=theta,  # Pre-defined theta
                cutoff_X=False,
                lambda_diff=LAMBDA_DIFF,
                flag_self_reporter=True,
                seed=seed,
                verbose=False,
            )

        lambda_k_GT = np.array([[LAMBDA_0, LAMBDA_0 + LAMBDA_DIFF]])
        beta_lambda = 10000 * np.ones((lambda_k_GT.shape))
        alpha_lambda = lambda_k_GT * beta_lambda

        model = vm.model.VimureModel(mutuality=True, verbose=False)
        with pytest.warns(None) as record:
            with suppress_stdout_stderr():
                model.fit(
                    gt_network.X,
                    K=K,
                    seed=seed,
                    theta_prior=(0.1, 0.1),
                    eta_prior=(0.5, 1),
                    alpha_lambda=alpha_lambda,
                    beta_lambda=beta_lambda,
                    num_realisations=2,
                    max_iter=21,
                    R=gt_network.R,
                )

        Y_true = gt_network.Y.toarray()[0].flatten()
        Y_rec = vm.utils.apply_rho_threshold(model, threshold=0.5)[0].flatten()

        assert np.allclose(f1_score(Y_true, Y_rec), 0.97, atol=1e-2)


class TestVimureWithReadData:
    """
    Tests the model with real data
    """

    def test_inform_nodes_reporters(self, karnataka_edgelist_vil1_money):
        df, nodes, reporters = karnataka_edgelist_vil1_money

        with pytest.warns(None) as record:
            net_obj = vm._io.read_from_edgelist(df, K=2, nodes=list(nodes), reporters=list(reporters))

        model = vm.model.VimureModel(mutuality=True, verbose=False)
        with suppress_stdout_stderr():
            model.fit(
                net_obj.X,
                K=2,
                seed=1,
                theta_prior=(0.1, 0.1),
                eta_prior=(0.5, 1),
                num_realisations=2,
                max_iter=21,
                R=net_obj.R,
            )


class TestInferredModel:
    @classmethod
    def setup_class(cls):
        cls.synth_net = vm.synthetic.StandardSBM(K=2)
        cls.synth_net._build_X(flag_self_reporter=True)

        cls.model = vm.model.VimureModel()
        with suppress_stdout_stderr():
            cls.model.fit(
                cls.synth_net.X,
                K=cls.synth_net.K,
                R=cls.synth_net.R,
                num_realisations=1,
                max_iter=500,
            )

        cls.model_mutuality = vm.model.VimureModel(mutuality=True)
        cls.model_mutuality.fit(
            cls.synth_net.X,
            K=cls.synth_net.K,
            R=cls.synth_net.R,
            num_realisations=1,
            max_iter=500,
        )

    def check_output(self, Y, model):
        assert Y.shape == (model.L, model.N, model.N)
        assert Y.sum() > 0

    def test_non_implemented_method(self):
        with pytest.raises(ValueError) as e_info:
            self.model.get_inferred_model(method="NotImplemented")
            assert (
                str(e_info.value)
                == '\'method\' should be one of "rho_max", "rho_mean","fixed_threshold", "heuristic_threshold".'
            )

    def test_fixed_threshold(self):
        with pytest.raises(ValueError) as e_info:
            self.model.get_inferred_model(method="fixed_threshold")
            assert (
                str(e_info.value)
                == 'For method="fixed_threshold", you must set the threshold to a value in [0,1].'
            )

        with pytest.raises(ValueError) as e_info:
            self.model.get_inferred_model(method="fixed_threshold", threshold=2)
            assert (
                str(e_info.value)
                == 'For method="fixed_threshold", you must set the threshold to a value in [0,1].'
            )

        Y = self.model.get_inferred_model(method="fixed_threshold", threshold=0.5)
        self.check_output(Y, self.model_mutuality)

    def test_rho_max(self):
        Y = self.model_mutuality.get_inferred_model(method="rho_max")
        self.check_output(Y, self.model_mutuality)

        Y = self.model.get_inferred_model(method="rho_max")
        self.check_output(Y, self.model)

    def test_rho_mean(self):
        Y = self.model_mutuality.get_inferred_model(method="rho_mean")
        self.check_output(Y, self.model_mutuality)

        Y = self.model.get_inferred_model(method="rho_mean")
        self.check_output(Y, self.model)

    def test_heuristic_threshold(self):
        Y = self.model_mutuality.get_inferred_model(
            method="heuristic_threshold"
        )
        self.check_output(Y, self.model_mutuality)

        Y = self.model.get_inferred_model(method="heuristic_threshold")
        self.check_output(Y, self.model)
    
    def test_sample_Y(self):
        N = 10
        Y = self.model.sample_inferred_model(N=N)
        
        assert len(Y) == N
        
        for y in Y:
            self.check_output(y, self.model_mutuality)


class TestVimureRealData:

    def test_internal_api(self, karnataka_edgelist_vil1_money):
    
        df, _, _ = karnataka_edgelist_vil1_money

        with pytest.warns(None) as record:
            net_obj = vm._io.read_from_edgelist(df, K=2)

        model = vm.model.VimureModel()

        with suppress_stdout_stderr():
            model.fit(net_obj.X, K=net_obj.K, R=net_obj.R)

        check_final_parameters(model, net_obj)

    def test_data_as_edgelist(self, karnataka_edgelist_vil1_money):
    
        df, _, _ = karnataka_edgelist_vil1_money

        # Run model directly with a pandas dataframe
        model = vm.model.VimureModel()
        with pytest.warns(None) as record:
            with suppress_stdout_stderr():
                model.fit(df, 
                        seed=1,
                        num_realisations=1,
                        max_iter=500)

        # Read in data as an edgelist to compare
        with pytest.warns(None) as record:
            net_obj = vm._io.read_from_edgelist(df, K=2)
        check_final_parameters(model, net_obj)

File: src/R/tests/testthat/test-model.R
require(testthat)
require(reticulate)

skip_if_no_vimure <- function() {
  have_vimure <- reticulate::py_module_available("vimure")
  if (!have_vimure)
    skip("Vimure not available for testing")
}

check_parameters <- function(model, synth_net) {
  expect_false(is.null(model$rho))
  expect_equal(dim(model$rho),
               c(synth_net$L, synth_net$N, synth_net$N, synth_net$K))
  expect_gt(sum(model$rho), 0)

  expect_false(is.null(model$gamma_shp))
  expect_equal(dim(model$gamma_shp), c(synth_net$L, synth_net$M))
  expect_gt(sum(model$gamma_shp), 0)

  expect_false(is.null(model$gamma_rte))
  expect_equal(dim(model$gamma_rte), c(synth_net$L, synth_net$M))
  expect_gt(sum(model$gamma_rte), 0)

  expect_false(is.null(model$phi_shp))
  expect_equal(dim(model$phi_shp), c(synth_net$L, synth_net$K))
  expect_gt(sum(model$phi_shp), 0)

  expect_false(is.null(model$phi_rte))
  expect_equal(dim(model$phi_rte), c(synth_net$L, synth_net$K))
  expect_gt(sum(model$phi_rte), 0)

  expect_false(is.null(model$nu_shp))
  expect_gt(sum(model$nu_shp), 0)

  expect_false(is.null(model$nu_rte))
  expect_gt(sum(model$nu_rte), 0)
}

check_extreme_scenarios <- function(exaggeration_type, f1_expected){
  eta <- 0.2
  theta_ratio <- 0.1
  seed <- 25
  K <- 2

  synth_net <- gm_crep(N=100, M=100, L=1, C=2,
                               K=K, avg_degree=5, sparsify=T, seed=seed, eta=eta
  )

  custom_theta <- build_custom_theta(
    synth_net,
    theta_ratio=theta_ratio,
    exaggeration_type=exaggeration_type,
    seed = seed
  )

  LAMBDA_0 <- 0.01
  LAMBDA_DIFF <- 0.99
  X <- build_X(synth_net, mutuality=eta, theta=custom_theta, cutoff_X=F, lambda_diff=LAMBDA_DIFF,
               flag_self_reporter=T, verbose=T
  )

  lambda_k_GT <- matrix(c(LAMBDA_0, LAMBDA_0 + LAMBDA_DIFF), ncol=K)
  beta_lambda <- matrix(10000 * rep(1, length(lambda_k_GT)), ncol=K)
  alpha_lambda <- lambda_k_GT * beta_lambda

  model <- vimure(synth_net$X, R=synth_net$R, K=K, seed=seed, mutuality=T,
                  alpha_lambda=alpha_lambda, beta_lambda=beta_lambda,
                  num_realisations=2, max_iter=21, verbose=F
  )

  Y_true <- as.vector(synth_net$Y$toarray())
  Y_rec <- as.integer(model$rho_f[,,,2] >= 0.5)
  tab <- table(Y_true, Y_rec)
  f1_score <- (tab[2,2])/(tab[2,2] + 0.5*(tab[1,2] + tab[2,1]))
  expect_lt(abs(f1_expected-f1_score), 1e-2)
}

check_output <- function(Y, model){
  expect_gt(sum(Y), 0)
  expect_equal(dim(Y), c(model$L, model$N, model$N))
}

test_that("Check reporters' mask", {
  skip_if_no_vimure()

  N <- 20
  nodes <- 1:N
  synth_net <- gm_standard_sbm(N = N, M = 20, L = 1, K = 3, C = 2,
                              avg_degree = 2, sparsify = FALSE)

  # specify some sort of mask
  # for example, only egos of synth_net are reporters
  adj_matrix <- synth_net$Y$toarray()[1, , ]
  reporters <- which(rowSums(adj_matrix) > 0)

  reporter_mask <- array(0,
                        dim = c(1, N, N, N),
                        dimnames = list("layer", nodes, nodes, nodes))

  reporter_mask[, , , reporters] <- "clearly_wrong"

  expect_error(vimure(adj_matrix, R = reporter_mask), 
               "R must be either entirely made of TRUE/FALSE or of 0s/1s.")

})

test_that("Check vimure model with standard_sbm", {
  skip_if_no_vimure()

  synth_net <- gm_standard_sbm(N=20, M=20, L=1, K=3, C=2, avg_degree=2, sparsify=F)
  X <- build_X(synth_net, flag_self_reporter=T)
  model <- vimure(synth_net$X, R=synth_net$R, K=synth_net$K, verbose=F)
  check_parameters(model, synth_net)
})

# Motivated by https://github.com/latentnetworks/vimure/issues/92
test_that("Check custom reporters' mask are enforced", {
  skip_if_no_vimure()

  L <- 1
  N <- 20
  nodes <- 1:N
  synth_net <- gm_standard_sbm(N = N, M = 20, L = L, K = 3, C = 2,
                               avg_degree = 2, sparsify = FALSE)

  X <- build_X(synth_net, flag_self_reporter = TRUE)

  # Essentially all 1:M are reporters
  all_reporters <- unique(synth_net$R$subs[[4]])

  reporter_mask <- array(0, dim = c(1, N, N, N),
                         dimnames = list("layer", 1:N, 1:N, 1:N))

  # I made it so that only egos are reporters
  for (l in 1:L) {
    for (i in 1:N) {
      for (j in 1:N) {
        for (m in seq_along(all_reporters)) {
          if (i == all_reporters[m]) {
            reporter_mask[l, i, j, m] <- 1
          }
        }
      }
    }
  }

  model <- vimure(synth_net$X,
                  R = reporter_mask,
                  K = synth_net$K,
                  verbose = TRUE)
  check_parameters(model, synth_net)

  # The object reporter_mask is already a native R matrix
  # Now I need to convert model$R from python to R

  # I will need numpy
  np <- import("numpy", convert = FALSE)

  # Initialise an empty 4D numpy array
  model_R <- np$zeros(shape = as.integer(c(L, N, N, N)))

  # Get the keys (a 4D list) of the python object model$R
  model_R_keys <- model$R$subs

  #For each key, assign the corresponding value to the numpy array
  for (i in seq_along(model_R_keys[[1]])) {
    model_R[model_R_keys[[1]][i], model_R_keys[[2]][i],
             model_R_keys[[3]][i], model_R_keys[[4]][i]] <- 1
  }

  model_R <- as.array(model_R)
  dimnames(model_R) <- list("layer", 1:N, 1:N, 1:N)

  expect_equal(model_R, reporter_mask)

})

test_that("Check vimure model in extreme scenarios of under reporting", {
  check_extreme_scenarios("under", 0.97)
})

test_that("Check vimure model in extreme scenarios of over reporting", {
  check_extreme_scenarios("over", 0.92)
})

test_that("Check vimure model for invalid inputs", {
  expect_error(vimure("a"), "invalid 'type'")
  expect_error(vimure(list(a = "a")), "invalid 'type'")
  expect_error(vimure(mtcars),
               "ValueError: Required columns not found in data frame")
  expect_error(vimure(1:10), "invalid 'type'")
})


test_that("Check inferred model", {
  synth_net <- gm_standard_sbm(N=20, M=20, sparsify = F, K=2)
  X <- build_X(synth_net, flag_self_reporter = T, cutoff_X = T)

  model <- vimure(synth_net$X, synth_net$R, num_realisations=1, max_iter=500, mutuality = T, verbose=F)
  expect_error(get_inferred_model(model, method = "NonImplemented"), "'method' should be one of")
  expect_error(get_inferred_model(model, method = "fixed_threshold"), 'you must set the threshold to a value')
  expect_error(get_inferred_model(model, method = "fixed_threshold", threshold=2), 'you must set the threshold to a value')

  OPTIONS <- c("rho_max", "rho_mean", "fixed_threshold", "heuristic_threshold")
  for(method in OPTIONS){
    Y_hat <- get_inferred_model(model, method = method, threshold = 0.5)
    check_output(Y_hat, model)
  }

  N <- 10L
  Y_hat <- sample_inferred_model(model, N=N)
  expect_equal(length(Y_hat), N)

  for(y in Y_hat){
    check_output(y, model)
  }
})
